text/0000-router-service.md

```diff
@@ -0,0 +1,434 @@
+
+Both events receive a single `transition` argument as described in the "Transition Object" section below, which explains the meaning of `from` and `to` in more detail.
+
+Redirection example:
+
+ 1. current route is A
+ 2. user initiates a transition to B
+ 3. routeWillChange fires `from` A `to` B.
+ 4. B redirects to C
+ 5. routeWillChange fires `from` A `to` C.
+ 6. routeDidChange fires `from` A `to` C.
+
+Abort example:
+
+ 1. current route is A
+ 2. user initiates a transition to B
+ 3. routeWillChange fires `from` A `to` B.
+ 4. in response to the previous routeWillChange event, the transition is aborted.
+ 5. routeWillChange fires `from` A `to` A.
+ 8. routeDidChange fires `from` A `to` A.
+
+Error example:
+
+ 1. current route is A
+ 2. user initiates a transition to B.index
+ 3. routeWillChange fires `from` A `to` B.
+ 4. B throws an exception, and the router discovers a "B-error" template.
+ 5. routeWillChange fires `from` A `to` B-error
+ 6. routeDidChange fires `from` A `to` B-error
+
+These are events, not extension hooks -- now that we are exposing a Service, it makes more sense to subscribe to its events than extend it.
+
+### New Properties
+
+`currentRoute`: an observable property. It is guaranteed to change whenever a route transition happens (even when that transition only changes parameters and doesn't change the active route). You should consider its value deeply immutable -- we will replace the whole structure whenever it changes. The value of `currentRoute` is a `RouteInfo` representing the current leaf route. `RouteInfo` is described below.
+
+`currentRouteName`:  a convenient alias for `currentRoute.name`.
+
+`currentURL`: provides the serialized string representing `currentRoute`.
+
+### Query Parameter Semantics
+
+Today, `queryParams` impose unnecessarily high cost because we cannot generate URLs or determine if a link is active without taking into account the default values of query parameters. Determining their default values is expensive, because it involves instantiating the corresponding controller, even in cases where we will never visit its route.
+
+Therefore, the `queryParams` argument to the new `urlFor`, `transitionTo`, `replaceWith`, and `isActive` methods defined in this document will behave differently.
+
+ - default values will not be stripped from generated URLs. For example, `urlFor('my-route', { sortBy: 'title' })` will always include `?sortBy=title`, whether or not `title` is the default value of `sortBy`.
+
+ - to explicitly unset a query parameter, you can pass the symbol `Ember.DEFAULT_VALUE` as its value. For example, `transitionTo('my-route', { sortBy: Ember.DEFAULT_VALUE })` will result in a URL that does not contain any `?sortBy=`.
```

Comment:
Rereading this now with a bit more context and explaining it for future travelers (read: me, six months from now).

This is to ensure that our default behavior of "hydrating unsupplied query params" continues to work. Without this we effectively end up doing `Object.assign({}, unsupplied, queryParams);` which would make `queryParams` only additive, which is wrong. We can't use values _other_ than this symbol as it is possible that it may be invalid.

1. I'm concerned about the name as `DEFAULT_VALUE` implies more than we likely mean to convey. I'd nominate `UNSET_VALUE`, `UNBIND_VALUE`, or `DELETE_VALUE` instead.
2. Can we stash this somewhere not on the `Ember` global?